org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InitHandler.java

```diff
@@ -223,6 +226,12 @@ public boolean belongsTo(Object family) {
 		};
 		job.setPriority(Job.BUILD);
 		job.setRule(ResourcesPlugin.getWorkspace().getRoot());
+		try {
+			// a workaround for https://github.com/redhat-developer/vscode-java/issues/2020
+			JobHelpers.waitForBuildJobs(2 * 60 * 1000); // 2 minutes
```

Comment:
wow that's super hacky. 
What happens if the projects genuinely needs more than 2min to build?

Reply:
We just wait org.eclipse.core.internal.events.AutoBuildJob.AutoBuildOffJob to finish. It certainly lasts less than 2 minutes. We set up autobuild to off at https://github.com/eclipse/eclipse.jdt.ls/blob/master/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/InitHandler.java#L193

Reply:
Forgive me but I don't see how that is supposed to work. Before the fix, we just scheduled the job, so it'd run right after AutoBuildOffJob finished, so ... same difference?

Reply:
The issue happens when the Initialize Workspace job runs before AutoBuildOffJob has been finished. It can happen sometimes. You can try to debug the JDT LS  startup.

Reply:
The commit https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/commit/?id=298e47435e19b1217e5593f3c29652720857f6e1 you shared used the following wait logic during changing autobuild flag, you should move the workaround logic to the utility `projectsManager.setAutoBuilding` as well.
![image](https://user-images.githubusercontent.com/14052197/125881997-002a7d52-56db-4128-81b1-6c38dc062627.png)

Reply:
Also you could modify the utility to `projectsManager.setAutoBuilding(boolean enable, boolean waitForBuildEvents)` to control whether to enable the wait. It looks like it's needed for AutoBuildOffJob, not sure for AutoBuildOnJob.

Reply:
I have tried it. It doesn't help. Java LS freezes at the line
```
Job.getJobManager().join(ResourcesPlugin.FAMILY_AUTO_BUILD, null);
```

See also https://github.com/eclipse/eclipse.jdt.ls/pull/1766#issuecomment-840411678

Reply:
https://github.com/eclipse/eclipse.jdt.ls/pull/1766#issuecomment-840411678  is different, where it skips the line `Job.getJobManager().join(InitHandler.JAVA_LS_INITIALIZATION_JOBS, null);` instead of freezing it there.

One part i‘m confused is why there is race condition between Initialize Workspace Job and AutoBuildOffJob. In other words, it's who triggered the AutoBuildOffJob. I feel we need more clarification about the root cause.

Reply:
> #1766 (comment) is different, where it skips the line Job.getJobManager().join(InitHandler.JAVA_LS_INITIALIZATION_JOBS, null); instead of freezing it there.

See 
> Changing to JobHelpers.waitForInitializeJobs(); works on my side. In waitForInitializeJobs() it directly get the initialization jobs no matter what status it is, I guess that's why it works.

and
https://github.com/eclipse/eclipse.jdt.ls/pull/1767/files

> One part i‘m confused is why there is race condition between Initialize Workspace Job and AutoBuildOffJob. In other words, it's who triggered the AutoBuildOffJob. I feel we need more clarification about the root cause.

JavaCore.start() -  https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/tree/org.eclipse.jdt.core/model/org/eclipse/jdt/core/JavaCore.java
The line 
```
JavaModelManager.getJavaModelManager().startup();
```

Reply:
I prefer to move the workaround code to the beginning (e.g. Line 189) of `triggerInitialization(Collection<IPath> roots)`.

Reply:
I have moved it.

Reply:
I'm still seeing the workaround code at the bottom. The new changes seem to be brought in from a rebase.

Reply:
Fixed.